# 1). Set Up the Environment

In [24]:
#import base libraries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [45]:
# import train dataset
google_review_data = pd.read_json("review-Vermont_10.json", lines=True)
advertisements_data = pd.read_json("advertisements.json")
irrelevant_data = pd.read_json("irrelevant.json")
rant_data = pd.read_json("rant_without_visit.json")

# 2). Exploratory Data Analysis

2a). Visualise the Google Review dataset...

In [46]:
print("\n----------FIRST FIVE INSTANCES IN GOOGLE REVIEW----------\n")
display(google_review_data.head())
print("\n----------GOOGLE REVIEW DATA INFORMATION----------\n")
display(google_review_data.info())
print("\n----------GOOGLE REVIEW DATA SUMMARY STATS----------\n")
display(google_review_data.describe())


----------FIRST FIVE INSTANCES IN GOOGLE REVIEW----------



user_id             name           time  rating  \
0  1.180269e+20    rebecca kerns  1620085852324       5   
1  1.015327e+20    Peter DeForge  1580309946474       5   
2  1.154041e+20    Chad Goulette  1605195974445       5   
3  1.047893e+20  Mark LaFountain  1593005848256       5   
4  1.089807e+20           Jeff R  1582059996120       5   

                                                text  pics  \
0      Always done right from wood stove to screens!  None   
1  A great company to work with.  Their sales and...  None   
2  Great place to do business with staff was grea...  None   
3  Awesome Customer service, quick response, and ...  None   
4  If you need a top quality job, by a group of p...  None   

                                                resp  \
0  {'time': 1620087641504, 'text': 'Good Evening,...   
1  {'time': 1580320228721, 'text': 'Good Afternoo...   
2  {'time': 1605195166792, 'text': 'Hi Chad!

Tha...   
3  {'time': 1593376422014, 'text': 'Mark, thank y...   
4  {'time': 1582063833737, 'text': 'Good Afternoo...   

                                 gmap_id  
0  0x89e02445cb9db457:0x37f42bff4edf7a43  
1  0x89e02445cb9db457:0x37f42bff4edf7a43  
2  0x89e02445cb9db457:0x37f42bff4edf7a43  
3  0x89e02445cb9db457:0x37f42bff4edf7a43  
4  0x89e02445cb9db457:0x37f42bff4edf7a43


----------GOOGLE REVIEW DATA INFORMATION----------

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 324725 entries, 0 to 324724
Data columns (total 8 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   user_id  324725 non-null  float64
 1   name     324725 non-null  object 
 2   time     324725 non-null  int64  
 3   rating   324725 non-null  int64  
 4   text     178168 non-null  object 
 5   pics     9651 non-null    object 
 6   resp     26838 non-null   object 
 7   gmap_id  324725 non-null  object 
dtypes: float64(1), int64(2), object(5)
memory usage: 19.8+ MB


None


----------GOOGLE REVIEW DATA SUMMARY STATS----------



user_id          time         rating
count  3.247250e+05  3.247250e+05  324725.000000
mean   1.091867e+20  1.548171e+12       4.314346
std    5.313166e+18  3.945905e+10       0.996339
min    1.000018e+20  1.182960e+12       1.000000
25%    1.045707e+20  1.523215e+12       4.000000
50%    1.091955e+20  1.550804e+12       5.000000
75%    1.137058e+20  1.573925e+12       5.000000
max    1.184442e+20  1.631106e+12       5.000000


Key Takeaways

- Columns with missing data in the train dataset: text, pics and resp
- Train dataset contains 324725 examples with 8 variables

2b). Visualise the advertisements, irrelevant, and rant_without_visit datasets...

In [47]:
print("\n----------FIRST FIVE INSTANCES IN ADVERTISEMENTS----------\n")
display(advertisements_data.head())
print("\n----------ADVERTISEMENTS DATA INFORMATION----------\n")
display(advertisements_data.info())
print("\n----------ADVERTISEMENTS DATA SUMMARY STATS----------\n")
display(advertisements_data.describe())

print("\n----------FIRST FIVE INSTANCES IN IRRELEVANT----------\n")
display(irrelevant_data.head())
print("\n----------IRRELEVANT DATA INFORMATION----------\n")
display(irrelevant_data.info())
print("\n----------IRRELEVANT DATA SUMMARY STATS----------\n")
display(irrelevant_data.describe())

print("\n----------FIRST FIVE INSTANCES IN RANT----------\n")
display(rant_data.head())
print("\n----------RANT DATA INFORMATION----------\n")
display(rant_data.info())
print("\n----------RANT DATA SUMMARY STATS----------\n")
display(rant_data.describe())


----------FIRST FIVE INSTANCES IN ADVERTISEMENTS----------



reviews
0  {'text': 'The food at Luigi's Pizzeria was dec...
1  {'text': 'I was at Oak Creek Dental for a clea...
2  {'text': 'The handyman at City Hardware helped...
3  {'text': 'Stopped by The Daily Grind Cafe for ...
4  {'text': 'The new tires I got from Pete's Auto...


----------ADVERTISEMENTS DATA INFORMATION----------

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101 entries, 0 to 100
Data columns (total 1 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   reviews  101 non-null    object
dtypes: object(1)
memory usage: 940.0+ bytes


None


----------ADVERTISEMENTS DATA SUMMARY STATS----------



reviews
count                                                 101
unique                                                101
top     {'text': 'The food at Luigi's Pizzeria was dec...
freq                                                    1


----------FIRST FIVE INSTANCES IN IRRELEVANT----------



reviews
0  {'text': 'My cat, Mr. Whiskers, learned how to...
1  {'text': 'The current administration's economi...
2  {'text': 'I prefer the taste of name-brand ket...
3  {'text': 'The weather this week has been atroc...
4  {'text': 'I finally finished that 5000-piece j...


----------IRRELEVANT DATA INFORMATION----------

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 1 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   reviews  100 non-null    object
dtypes: object(1)
memory usage: 932.0+ bytes


None


----------IRRELEVANT DATA SUMMARY STATS----------



reviews
count                                                 100
unique                                                100
top     {'text': 'My cat, Mr. Whiskers, learned how to...
freq                                                    1


----------FIRST FIVE INSTANCES IN RANT----------



reviews
0  {'text': 'Absolutely the worst place ever. I w...
1  {'text': 'I heard from a friend of a friend ab...
2  {'text': 'After reading some things online abo...
3  {'text': 'The service is notoriously awful. Ev...
4  {'text': 'I drove by this place and just from ...


----------RANT DATA INFORMATION----------

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 1 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   reviews  100 non-null    object
dtypes: object(1)
memory usage: 932.0+ bytes


None


----------RANT DATA SUMMARY STATS----------



reviews
count                                                 100
unique                                                100
top     {'text': 'Absolutely the worst place ever. I w...
freq                                                    1


Key Takeaways

- Advertisements, Irrelevant and Rant datasets contain 101, 100, and 100 examples respectively.

# 4). Preprocessing and Modeling 

In [ ]:
import ollama
import re
from sklearn.metrics import precision_score, recall_score, f1_score

4a). Preprocessing

In [64]:
genuine_texts = [item for item in google_review_data[:200]["text"] if item is not None]
# According to deepseek r1, these reviews are all genuine

ad_reviews_list = advertisements_data["reviews"]
# Extract the 'text' from each dictionary in the list
ad_texts = [review["text"] for review in ad_reviews_list]

irre_reviews_list = irrelevant_data["reviews"]
# Extract the 'text' from each dictionary in the list
irre_texts = [review["text"] for review in irre_reviews_list]

rant_reviews_list = rant_data["reviews"]
# Extract the 'text' from each dictionary in the list
rant_texts = [review["text"] for review in rant_reviews_list]

# Construct our evaluation dataset
columns=['text', 'ad_score', 'irre_score', 'rant_score']
data = []
for text in genuine_texts[:80]:
    data.append([text, 0, 0, 0])
for text in ad_texts[:80]:
    data.append([text, 1, 0, 0])
for text in irre_texts[:80]:
    data.append([text, 0, 1, 0])
for text in rant_texts[:80]:
    data.append([text, 0, 0, 1])
evaluation_df = pd.DataFrame(data, columns=columns).sample(frac=1.0, random_state=42)
X, y = evaluation_df["text"], evaluation_df.drop(columns=["text"])

# retrieve examples for few-shot learning
print("genuine review example:", genuine_texts[81])
print("advertisement review example:", ad_texts[81])
print("irrelevant review example:", irre_texts[81])
print("rant without visit review example:", rant_texts[81])

# Visualise Evaluation Set
print("\n----------FIRST FIVE INSTANCES IN EVALUATION SET----------\n")
display(evaluation_df.head())
print("\n----------EVALUATION SET DATA INFORMATION----------\n")
display(evaluation_df.info())
print("\n----------EVALUATION SET DATA SUMMARY STATS----------\n")
display(evaluation_df.describe())

genuine review example: Great place to do business with staff was great,thank u!
advertisement review example: The new burger joint is a hit. I could only enjoy it because my heartburn is gone thanks to AcidFix tablets. Get a free trial bottle at GutHealthNow.net (just pay s&h). You can eat anything without worry!
irrelevant review example: I think reality TV has peaked and is now in a steady decline. The golden age is over.
rant without visit review example: I don't know how they sleep at night knowing how badly they operate. The lack of care is astounding. Everything is a mess.

----------FIRST FIVE INSTANCES IN EVALUATION SET----------



text  ad_score  irre_score  \
167  The new season of my favorite baking show is a...         0           1   
230  My neighbor's lawn is looking exceptionally gr...         0           1   
25   These people are really good. They sent Mike C...         0           0   
63   Sandwich was awesome!   Thick and flavorful.  ...         0           0   
9         Love Carrs....always have very nice people!!         0           0   

     rant_score  
167           0  
230           0  
25            0  
63            0  
9             0


----------EVALUATION SET DATA INFORMATION----------

<class 'pandas.core.frame.DataFrame'>
Index: 320 entries, 167 to 102
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   text        320 non-null    object
 1   ad_score    320 non-null    int64 
 2   irre_score  320 non-null    int64 
 3   rant_score  320 non-null    int64 
dtypes: int64(3), object(1)
memory usage: 12.5+ KB


None


----------EVALUATION SET DATA SUMMARY STATS----------



ad_score  irre_score  rant_score
count  320.000000  320.000000  320.000000
mean     0.250000    0.250000    0.250000
std      0.433691    0.433691    0.433691
min      0.000000    0.000000    0.000000
25%      0.000000    0.000000    0.000000
50%      0.000000    0.000000    0.000000
75%      0.250000    0.250000    0.250000
max      1.000000    1.000000    1.000000

4b). Evaluating Model: `deepseek-r1:14b`

In [66]:
# Prompting and Prediction
batch_size = 8

i = 0
pred_data = []
columns = ['ad_score', 'irre_score', 'rant_score']
y_pred = pd.DataFrame(pred_data, columns=columns)

# Start with the initial prompt with few-shot learning
content = """
You are an expert review classifier. Your task is to analyze user reviews and assign a confidence level for three specific categories: "advertisement", "irrelevant", and "rant without visit".

**Categories & Examples:**
-   **advertisement**: "Best pizza! Visit www.pizzapromo.com for discounts!"
-   **irrelevant**: "I love my new phone, but this place is too noisy."
-   **rant without visit**: "Never been here, but i heard it's terrible."

**Input:**
A list of reviews.

**Task:**
For the provided reviews, analyze the text to determine a confidence level (Low, or High) for each category.

**Output Instructions:**
- **DO NOT** include any other text, explanations, or commentary before or after the JSON.
- **DO NOT** include a `reviews` or any other key for the main JSON array.
- The output must be a single JSON array `[]` containing one JSON object `{}` for each review. All reviews must be analysed.
- Each object must have exactly three keys: `"advertisement"`, `"irrelevant"`, and `"rant without visit"`.
- Each value must be one of the strings: `"Low"` or `"High"`.
- The format must be identical to the example output.

**Example Input:**
Great place to do business with staff was great,thank u!", "The new burger joint is a hit. I could only enjoy it because my heartburn is gone thanks to AcidFix tablets. Get a free trial bottle at GutHealthNow.net (just pay s&h). You can eat anything without worry!", "I think reality TV has peaked and is now in a steady decline. The golden age is over.", "I don't know how they sleep at night knowing how badly they operate. The lack of care is astounding. Everything is a mess."

**Example Output:**
```json
[
  {
    "advertisement": "Low",
    "irrelevant": "Low",
    "rant without visit": "Low"
  },
  {
    "advertisement": "High",
    "irrelevant": "Low",
    "rant without visit": "Low"
  },
  {
    "advertisement": "Low",
    "irrelevant": "High",
    "rant without visit": "Low"
  },
  {
    "advertisement": "Low",
    "irrelevant": "Low",
    "rant without visit": "High"
  }
]
"""
messages=[
    {'role': 'user', 'content': content}
]

# 2. Make the first call
response1 = ollama.chat(model='deepseek-r1:14b', messages=messages) # requires the model to be installed locally in ollama
print(response1['message']['content'])

# 3. Append the model's response to the message history
messages.append({'role': 'assistant', 'content': response1['message']['content']})

while i < len(evaluation_df): #len(evaluation_df)
  print(f"Processing reviews {str(i+1)} to {str(i+batch_size)} out of {len(evaluation_df)}")
  # 4. Add the next user prompt to test reviews
  content2 = ", ".join([f'"{item}"' for item in X[i:i+batch_size]])
  messages.append({'role': 'user', 'content': content2})

  # 5. Make the second call, passing the entire updated message list
  response2 = ollama.chat(model='deepseek-r1:14b', messages=messages)
  messages.pop() # remove the previous message

  index = response2['message']['content'].find("</think>") + len("</think>")
  actual_output = response2['message']['content'][index:]
  print(actual_output)

  raw_pred = []
  words_to_find = ["advertisement", "irrelevant", r"rant( |_)without( |_)visit"]
  for word in words_to_find:
      word_count = 0
      pattern = r'(\"|\*\*|)' + word + r'(\"|\*\*|):'
      for match in re.finditer(pattern, actual_output, re.IGNORECASE):
          print(f"Found '{match.group()}' at index {match.start()}")
          print(actual_output[match.end():match.end()+10].lower())
          if "low" in actual_output[match.end():match.end()+10].lower():
              raw_pred.append(0)
          if "high" in actual_output[match.end():match.end()+10].lower():
              raw_pred.append(1)
          word_count += 1
      if word_count != batch_size:
          print(f"word_count expected {batch_size} but got {word_count}")
          continue

  if len(raw_pred) != 3*batch_size:
      print(f"length incorrect, expected {3*batch_size} but got {len(raw_pred)}")
      continue

  columns = ['ad_score', 'irre_score', 'rant_score']
  for j in range(batch_size):
      y_pred.loc[len(y_pred)] = [raw_pred[j], raw_pred[j+batch_size], raw_pred[j+2*batch_size]]

  i += batch_size

<think>
Okay, I need to analyze each review and assign confidence levels for advertisement, irrelevant, and rant without visit categories. Let's take it step by step.

First Review: "Great place to do business with staff was great,thank u!" 

- Advertisement: No mention of any website or product, so Low.
- Irrelevant: It talks about the place being good, but not unrelated. So probably Low.
- Rant without visit: They visited and had a positive experience, so no rant without visit. Low.

Second Review: "The new burger joint is a hit. I could only enjoy it because my heartburn is gone thanks to AcidFix tablets. Get a free trial bottle at GutHealthNow.net (just pay s&h). You can eat anything without worry!"

- Advertisement: They mention AcidFix tablets and provide a website, so High.
- Irrelevant: It's related to the burger joint, not irrelevant. So Low.
- Rant without visit: No negative ranting; it's positive. So Low.

Third Review: "I think reality TV has peaked and is now in a steady d

In [67]:
# Results

y_truth = y

for category in y_truth.columns:
    print(f"--- Metrics for {category} ---")

    true_labels = y_truth[category]
    pred_labels = y_pred[category]

    precision = precision_score(true_labels, pred_labels)
    recall = recall_score(true_labels, pred_labels)
    f1 = f1_score(true_labels, pred_labels)

    print(f"Precision: {precision:.2f}")
    print(f"Recall: {recall:.2f}")
    print(f"F1-Score: {f1:.2f}\n")



--- Metrics for ad_score ---
Precision: 0.96
Recall: 0.99
F1-Score: 0.98

--- Metrics for irre_score ---
Precision: 0.70
Recall: 0.86
F1-Score: 0.77

--- Metrics for rant_score ---
Precision: 0.85
Recall: 0.72
F1-Score: 0.78

